crypto/src/lib.rs

```diff
@@ -274,89 +299,189 @@ impl From<KeyPair> for (PublicKey, PrivateKey) {
+    /// Extracts the raw bytes from public key, copying the payload.
+    ///
+    /// `into_raw()` without copying is not provided because underlying crypto
+    /// libraries do not provide move functionality.
+    pub fn to_raw(&self) -> (Algorithm, Vec<u8>) {
+        (self.algorithm(), self.payload())
     }
 
     /// Key payload
-    // TODO: Derive with getset once FFI impl is fixed
-    pub fn payload(&self) -> &[u8] {
-        self.payload.as_ref()
-    }
-
-    #[cfg(feature = "std")]
-    fn try_from_private(private_key: PrivateKey) -> Result<PublicKey, Error> {
-        let digest_function = private_key.digest_function();
-        let key_gen_option = Some(KeyGenOption::FromPrivateKey(private_key));
-
-        let (public_key, _) = match digest_function {
-            Algorithm::Ed25519 => signature::ed25519::Ed25519Sha512::keypair(key_gen_option),
-            Algorithm::Secp256k1 => {
-                signature::secp256k1::EcdsaSecp256k1Sha256::keypair(key_gen_option)
-            }
-            Algorithm::BlsNormal => signature::bls::BlsNormal::keypair(key_gen_option),
-            Algorithm::BlsSmall => signature::bls::BlsSmall::keypair(key_gen_option),
-        }?;
+    fn payload(&self) -> Vec<u8> {
+        use w3f_bls::SerializableToBytes as _;
 
-        Ok(public_key)
+        match self {
+            PublicKey::Ed25519(key) => key.as_bytes().to_vec(),
+            PublicKey::Secp256k1(key) => key.to_sec1_bytes().to_vec(),
+            PublicKey::BlsNormal(key) => key.to_bytes(),
+            PublicKey::BlsSmall(key) => key.to_bytes(),
+        }
     }
 
-    /// Construct `PrivateKey` from hex encoded string
-    ///
+    /// Construct [`PublicKey`] from hex encoded string
     /// # Errors
     ///
     /// - If the given payload is not hex encoded
     /// - If the given payload is not a valid private key
-    #[cfg(feature = "std")]
-    pub fn from_hex(digest_function: Algorithm, payload: &str) -> Result<Self, Error> {
+    pub fn from_hex(digest_function: Algorithm, payload: &str) -> Result<Self, ParseError> {
         let payload = hex_decode(payload)?;
-        let payload = ConstVec::new(payload);
 
-        // NOTE: PrivateKey does some validation by generating a public key from the provided bytes
-        // we can't really do this for PublicKey
-        // this can be solved if the keys used here would be actually aware of the underlying crypto primitive types
-        // instead of just being raw bytes
-        Ok(Self {
-            digest_function,
-            payload,
+        Self::from_raw(digest_function, &payload)
+    }
+
+    /// Get the digital signature algorithm of the public key
+    pub fn algorithm(&self) -> Algorithm {
+        match self {
+            Self::Ed25519(_) => Algorithm::Ed25519,
+            Self::Secp256k1(_) => Algorithm::Secp256k1,
+            Self::BlsNormal(_) => Algorithm::BlsNormal,
+            Self::BlsSmall(_) => Algorithm::BlsSmall,
+        }
+    }
```

Comment:
this has become quite useless, since the user can use `match` or `matches`. I would remove it

Reply:
but we can leave it

Reply:
Most of the time we just need any `Algorithm` value to do further logic, not just check exact value